state/prunning_service.go

```diff
@@ -0,0 +1,228 @@
+// retain height for blocks or ABCI block results.
+// It returns the height which will be used as the retain height and an error message in case
+// the retain height cannot be set to the desired value (if the data was already pruned, if a lower
+// retain height was already set)
+func (p *Pruner) SetPruningHeight(retainHeightInfo RetainHeightInfo) (retainHeight int64, err error) {
+
+	// TODO we need to lock retrieval of base and height
+	if retainHeightInfo.Height <= 0 || retainHeightInfo.Height < p.bs.Base() || retainHeightInfo.Height > p.bs.Height() {
+		return 0, ErrInvalidHeightValue
+	}
+
+	var currentRetainHeight int64
+	switch retainHeightInfo.Requester {
+	case AppRequester:
+		currentRetainHeight, err = p.stateStore.GetApplicationRetainHeight()
+		if err != nil {
+			if err == ErrKeyNotFound {
+				err = p.stateStore.SaveApplicationRetainHeight(retainHeightInfo.Height)
+				return retainHeightInfo.Height, err
+			}
+			return 0, err
+		}
+		if currentRetainHeight > retainHeightInfo.Height {
+			return currentRetainHeight, errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+		}
+		err = p.stateStore.SaveApplicationRetainHeight(retainHeightInfo.Height)
+		return retainHeightInfo.Height, err
+	case DataCompanionRequester:
+		currentRetainHeight, err = p.stateStore.GetDataCompanionRetainHeight()
+		if err != nil {
+			if err == ErrKeyNotFound {
+				err = p.stateStore.SaveDataCompanionRetainHeight(retainHeightInfo.Height)
+				return retainHeightInfo.Height, err
+			}
+			return 0, err
+		}
+		if currentRetainHeight > retainHeightInfo.Height {
+			return currentRetainHeight, errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+		}
+		err = p.stateStore.SaveDataCompanionRetainHeight(retainHeightInfo.Height)
+		return retainHeightInfo.Height, err
+	case ABCIResRetainHeightRequester:
+		currentRetainHeight, err = p.stateStore.GetABCIResRetainHeight()
+		if err != nil {
+			if err == ErrKeyNotFound {
+				err = p.stateStore.SaveABCIResRetainHeight(retainHeightInfo.Height)
+				return retainHeightInfo.Height, err
+			}
+			return 0, err
+		}
+		if currentRetainHeight > retainHeightInfo.Height {
+			return currentRetainHeight, errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+		}
+		err = p.stateStore.SaveABCIResRetainHeight(retainHeightInfo.Height)
+		return retainHeightInfo.Height, err
+	}
+
+	return 0, nil
+}
+func (p *Pruner) pruningRoutine() {
+	lastHeightPruned := int64(0)
+	lastABCIResPrunedHeight := int64(0)
+	for {
+		select {
+		case <-p.pruningTicker.C:
+			retainHeight := p.findMinRetainHeight()
+			if retainHeight != lastHeightPruned {
+				pruned, evRetainHeight, err := p.pruneBlocks(retainHeight)
+				p.logger.Info(fmt.Sprintf("Pruned %d, evidence retain height is %d , err %s", pruned, evRetainHeight, err))
```

Comment:
We should avoid using `fmt.Sprintf` in log messages. Rather:

```suggestion
				if err != nil {
					p.logger.Error("Failed to prune blocks", "err", err)
				} else {
					p.logger.Debug("Pruned block(s)", "height", pruned, "evidenceRetainHeight", evRetainHeight)
				}
```

The error should most likely be logged at a different level to the info regarding which heights were pruned. We should be very careful with using `Info` level logging, since we may end up polluting operators' logs.

If we log anything at `Info` level during pruning operations, it should exclusively be a summary and only when something was actually pruned (I imagine this would be useful for operators to see to know that pruning's working).